zcash_transparent/src/builder.rs

```diff
@@ -155,32 +155,57 @@ fn construct_script_sig(
     ])
 }
 
+/// Information required to spend a transparent coin; the secp256k public key when spending a p2pkh
+/// output, or the redeem script when spending a p2sh output.
 #[cfg(feature = "transparent-inputs")]
 #[derive(Debug, Clone)]
-enum InputKind {
+pub enum SpendInfo {
     P2pkh { pubkey: secp256k1::PublicKey },
     P2sh { redeem_script: script::FromChain },
 }
 
-// TODO: This feature gate can be removed.
+/// Information required for adding a transparent input to the builder and computing fees: the
+/// outpoint referencing the coin to be spent, the original coin data, and the [`SpendInfo`]
+/// required to spend the output.
 #[cfg(feature = "transparent-inputs")]
 #[derive(Debug, Clone)]
 pub struct TransparentInputInfo {
-    kind: InputKind,
     utxo: OutPoint,
     coin: TxOut,
+    spend_info: SpendInfo,
 }
 
 #[cfg(feature = "transparent-inputs")]
 impl TransparentInputInfo {
+    /// Constructs a [`TransparentInputInfo`] value from its constituent parts.
+    pub fn new(utxo: OutPoint, coin: TxOut, spend_info: SpendInfo) -> Self {
```

Comment:
I dislike exposing a constructor for `TransparentInputInfo` that doesn't enforce the correctness constraints that `Builder::add_input` and `Builder::add_p2sh_input` do. Instead, we should move the correctness checks from those methods into here, and then have those existing methods be thin wrappers around this one.

We could also change `Builder::add_input` to take `TransparentInputInfo` as its argument, and rename the current `Builder::add_input` to `Builder::add_p2pkh_input` (so it's explicitly a convenience function). That would then remove the need for the `TransparentInputInfo::into_parts` method.